File: maven-plugin/src/main/java/GoalBehavior.java
import java.util.ArrayList;
import java.util.List;

/**
 * Data class representing the behavior and requirements of a Maven goal.
 * Used to replace hardcoded goal classification logic with dynamic analysis.
 */
public class GoalBehavior {
    private boolean processesSources = false;
    private boolean testRelated = false;
    private boolean needsResources = false;
    private List<String> sourcePaths = new ArrayList<>();
    private List<String> resourcePaths = new ArrayList<>();
    private List<String> outputPaths = new ArrayList<>();
    
    public boolean processesSources() {
        return processesSources;
    }
    
    public void setProcessesSources(boolean processesSources) {
        this.processesSources = processesSources;
    }
    
    public boolean isTestRelated() {
        return testRelated;
    }
    
    public void setTestRelated(boolean testRelated) {
        this.testRelated = testRelated;
    }
    
    public boolean needsResources() {
        return needsResources;
    }
    
    public void setNeedsResources(boolean needsResources) {
        this.needsResources = needsResources;
    }
    
    public List<String> getSourcePaths() {
        return sourcePaths;
    }
    
    public void setSourcePaths(List<String> sourcePaths) {
        this.sourcePaths = new ArrayList<>(sourcePaths);
    }
    
    public List<String> getResourcePaths() {
        return resourcePaths;
    }
    
    public void setResourcePaths(List<String> resourcePaths) {
        this.resourcePaths = new ArrayList<>(resourcePaths);
    }
    
    public List<String> getOutputPaths() {
        return outputPaths;
    }
    
    public void setOutputPaths(List<String> outputPaths) {
        this.outputPaths = new ArrayList<>(outputPaths);
    }
    
    /**
     * Merge this behavior with another, taking the union of all capabilities.
     */
    public GoalBehavior merge(GoalBehavior other) {
        GoalBehavior merged = new GoalBehavior();
        merged.processesSources = this.processesSources || other.processesSources;
        merged.testRelated = this.testRelated || other.testRelated;
        merged.needsResources = this.needsResources || other.needsResources;
        
        merged.sourcePaths.addAll(this.sourcePaths);
        merged.sourcePaths.addAll(other.sourcePaths);
        
        merged.resourcePaths.addAll(this.resourcePaths);
        merged.resourcePaths.addAll(other.resourcePaths);
        
        merged.outputPaths.addAll(this.outputPaths);
        merged.outputPaths.addAll(other.outputPaths);
        
        return merged;
    }
    
    /**
     * Check if this behavior has any defined capabilities.
     */
    public boolean hasAnyBehavior() {
        return processesSources || testRelated || needsResources || 
               !sourcePaths.isEmpty() || !resourcePaths.isEmpty() || !outputPaths.isEmpty();
    }
}

File: maven-plugin/src/main/kotlin/GoalBehavior.kt
/**
 * Data class representing the behavior and requirements of a Maven goal.
 * Used to replace hardcoded goal classification logic with dynamic analysis.
 */
data class GoalBehavior(
    private var _processesSources: Boolean = false,
    private var _testRelated: Boolean = false,
    private var _needsResources: Boolean = false,
    private var _sourcePaths: MutableList<String> = mutableListOf(),
    private var _resourcePaths: MutableList<String> = mutableListOf(),
    private var _outputPaths: MutableList<String> = mutableListOf()
) {
    // Java-style getters for compatibility
    fun processesSources(): Boolean = _processesSources
    fun isTestRelated(): Boolean = _testRelated
    fun needsResources(): Boolean = _needsResources
    fun getSourcePaths(): MutableList<String> = _sourcePaths
    fun getResourcePaths(): MutableList<String> = _resourcePaths
    fun getOutputPaths(): MutableList<String> = _outputPaths
    
    // Java-style setters for compatibility
    fun setProcessesSources(value: Boolean) { _processesSources = value }
    fun setTestRelated(value: Boolean) { _testRelated = value }
    fun setNeedsResources(value: Boolean) { _needsResources = value }
    fun setSourcePaths(paths: MutableList<String>) { _sourcePaths = paths.toMutableList() }
    fun setResourcePaths(paths: MutableList<String>) { _resourcePaths = paths.toMutableList() }
    fun setOutputPaths(paths: MutableList<String>) { _outputPaths = paths.toMutableList() }
    /**
     * Merge this behavior with another, taking the union of all capabilities.
     */
    fun merge(other: GoalBehavior): GoalBehavior {
        return GoalBehavior(
            _processesSources = this._processesSources || other._processesSources,
            _testRelated = this._testRelated || other._testRelated,
            _needsResources = this._needsResources || other._needsResources,
            _sourcePaths = (this._sourcePaths + other._sourcePaths).toMutableList(),
            _resourcePaths = (this._resourcePaths + other._resourcePaths).toMutableList(),
            _outputPaths = (this._outputPaths + other._outputPaths).toMutableList()
        )
    }
    
    /**
     * Check if this behavior has any defined capabilities.
     */
    fun hasAnyBehavior(): Boolean {
        return _processesSources || _testRelated || _needsResources || 
               _sourcePaths.isNotEmpty() || _resourcePaths.isNotEmpty() || _outputPaths.isNotEmpty()
    }
}